pywr-core/src/solvers/builder.rs

```diff
@@ -397,22 +433,29 @@ where
     ) -> Result<(), PywrError> {
         let dt = timestep.days();
 
-        for (row_id, node) in self.node_constraints_row_ids.iter().zip(network.nodes().deref()) {
-            let (lb, ub): (f64, f64) = match node.get_current_flow_bounds(network, state) {
-                Ok(bnds) => bnds,
-                Err(PywrError::FlowConstraintsUndefined) => {
-                    // Must be a storage node
-                    let (avail, missing) = match node.get_current_available_volume_bounds(state) {
-                        Ok(bnds) => bnds,
-                        Err(e) => return Err(e),
-                    };
-
-                    (-avail / dt, missing / dt)
-                }
-                Err(e) => return Err(e),
+        for row in self.node_constraints_row_ids.iter() {
+            let node = network.get_node(&row.node_idx)?;
+            let (lb, ub): (f64, f64) = match node.get_bounds(network, state)? {
+                NodeBounds::Flow(bounds) => (bounds.min_flow, bounds.max_flow),
+                NodeBounds::Volume(bounds) => (-bounds.available / dt, bounds.missing / dt),
             };
 
-            self.builder.apply_row_bounds(*row_id, lb, ub);
+            match row.row_type {
+                NodeRowType::Continuous => {
+                    // Regular node constraint
+                    self.builder.apply_row_bounds(row.row_id.to_usize().unwrap(), lb, ub);
+                }
+                NodeRowType::Binary { bin_col_id } => {
+                    if !lb.is_zero() {
+                        panic!("Binary node constraint with non-zero lower bounds not implemented!");
+                    }
+                    // Update the coefficients for the binary column to be the upper bound
+                    self.builder.coefficients_to_update.push((row.row_id, bin_col_id, -ub));
+
+                    // This row is upper bounded to zero
+                    self.builder.apply_row_bounds(row.row_id.to_usize().unwrap(), FMIN, 0.0);
```

Comment:
are these bounds not already set in `create_node_constraints` and applied by the builder?

Reply:
Yes, they are.